core/touch_gesture.js

```diff
@@ -0,0 +1,287 @@
+ * @param {!Event} e A touch start, or pointer down event.
+ * @package
+ */
+Blockly.TouchGesture.prototype.handleTouchStart = function(e) {
+  var pointerId = Blockly.Touch.getTouchIdentifierFromEvent(e);
+  // store the pointerId in the current list of pointers
+  this.cachedPoints_[pointerId] = this.getTouchPoint(e);
+  var pointers = Object.keys(this.cachedPoints_);
+  // If two pointers are down, check for pinch gestures
+  if (pointers.length == 2) {
+    var point0 = this.cachedPoints_[pointers[0]];
+    var point1 = this.cachedPoints_[pointers[1]];
+    this.startDistance_ = goog.math.Coordinate.distance(point0, point1);
+    this.isMultiTouch_ = true;
+  }
+  e.preventDefault();
+};
+
+/**
+ * Handle a touch move or pointer move event and zoom in/out if two pointers are on the screen.
+ * @param {!Event} e A touch move, or pointer move event.
+ * @package
+ */
+Blockly.TouchGesture.prototype.handleTouchMove = function(e) {
+  var pointerId = Blockly.Touch.getTouchIdentifierFromEvent(e);
+  // Update the cache
+  this.cachedPoints_[pointerId] = this.getTouchPoint(e);
+
+  var pointers = Object.keys(this.cachedPoints_);
+  // If two pointers are down, check for pinch gestures
+  if (pointers.length == 2) {
+    // Calculate the distance between the two pointers
+    var point0 = this.cachedPoints_[pointers[0]];
+    var point1 = this.cachedPoints_[pointers[1]];
+    var moveDistance = goog.math.Coordinate.distance(point0, point1);
+    var startDistance = this.startDistance_;
+    var scale = this.touchScale_ = moveDistance / startDistance;
+    
+    if (this.previousScale_ > 0 && this.previousScale_ < Infinity) {
+      var gestureScale = scale - this.previousScale_;
+      var delta = gestureScale > 0 ? gestureScale * 5 : gestureScale * 6;
+      var workspace = this.startWorkspace_;
+      var position = Blockly.utils.mouseToSvg(e, workspace.getParentSvg(), workspace.getInverseScreenCTM());
+      workspace.zoom(position.x, position.y, delta);
+    }
+    this.previousScale_ = scale;
+  }
+  e.preventDefault();
+};
+
+/**
+ * Handle a touch end or pointer end event and end the gesture.
+ * @param {!Event} e A touch end, or pointer end event.
+ * @package
+ */
+Blockly.TouchGesture.prototype.handleTouchEnd = function(e) {
+  var pointerId = Blockly.Touch.getTouchIdentifierFromEvent(e);
+  if (this.cachedPoints_[pointerId]) {
+    delete this.cachedPoints_[pointerId];
+  }
+  if (Object.keys(this.cachedPoints_).length < 2) {
+    this.cachedPoints_ = {};
+    this.previousScale_ = 0;
+  }
+};
+
+/**
+ * Helper function returning the current touch point coordinate.
+ * @param {!Event} e A touch or pointer event.
+ * @return {goog.math.Coordinate} the current touch point coordinate
+ * @package
+ */
+Blockly.TouchGesture.prototype.getTouchPoint = function(e) {
+  if (!this.startWorkspace_) {
+    return null;
+  }
+  var metrics = this.metrics_ || (this.metrics_ = this.startWorkspace_.getMetrics());
```

Comment:
Why do you need the metrics?  All of the work that you're doing with these pointer locations involves deltas, so you shouldn't need the workspace info.  There is on call to Blockly.utils.mouseToSvg (line 248), but that uses the event directly and not the stored coordinate.

Reply:
I think you might be right, since the touch point is all relative, the distance can be calculated irrespective of the workspace's absolute position. I think I needed the metrics at some point for the zoom, but as you mentioned I'm  using mouseToSvg above instead.